eras/conway/impl/src/Cardano/Ledger/Conway/Rules/VDel.hs

```diff
@@ -12,44 +13,60 @@
 module Cardano.Ledger.Conway.Rules.VDel (
   ConwayVDEL,
   ConwayVDelEvent (..),
-  VDelEnv (..),
   ConwayVDelPredFailure,
 )
 where
 
 import Cardano.Ledger.BaseTypes (
   ShelleyBase,
  )
-import Cardano.Ledger.CertState (VState)
+import Cardano.Ledger.CertState (VState (VState, vsCommitteeHotKeys, vsDReps))
+import Cardano.Ledger.Coin (Coin)
 import Cardano.Ledger.Conway.Era (ConwayVDEL)
-import Cardano.Ledger.Conway.TxCert (ConwayCommitteeCert)
-import Cardano.Ledger.Core (Era (EraCrypto), EraRule)
+import Cardano.Ledger.Conway.TxCert (ConwayCommitteeCert (..))
+import Cardano.Ledger.Core (Era (EraCrypto), EraPParams, EraRule, PParams)
+import Cardano.Ledger.Credential (Credential)
+import Cardano.Ledger.Keys (KeyHash, KeyRole (CommitteeColdKey, Voting))
 import Control.DeepSeq (NFData)
-import Control.State.Transition (
+import Control.State.Transition.Extended (
   BaseM,
   Environment,
   Event,
   PredicateFailure,
   STS,
   Signal,
   State,
+  TRC (TRC),
+  TransitionRule,
+  judgmentContext,
   transitionRules,
+  (?!),
  )
+import qualified Data.Map.Strict as Map
+import Data.Maybe (isNothing)
+import qualified Data.Set as Set
 import GHC.Generics (Generic)
 import NoThunks.Class (NoThunks (..))
 
 data ConwayVDelPredFailure era
-  deriving (Show, Eq, Generic, NoThunks, NFData)
+  = ConwayDRepAlreadyRegisteredVDEL !(Credential 'Voting (EraCrypto era))
+  | ConwayDRepNotRegisteredVDEL !(Credential 'Voting (EraCrypto era))
+  | ConwayDRepIncorrectDepositVDEL !Coin
+  | ConwayCommitteeHasResignedVDEL !(KeyHash 'CommitteeColdKey (EraCrypto era))
+  | ConwayCommitteeNotRegisteredVDEL !(KeyHash 'CommitteeColdKey (EraCrypto era))
```

Comment:
You might wanna rebase on this PR and add Enc/DecCBOR instances for this type
https://github.com/input-output-hk/cardano-ledger/pull/3477

Soon above PR will get merged and you'll have to do it anyways, so it is just a heads up

Reply:
Oh yes, I saw @teodanciu PR just now. Got it! I will rebase on top of it and update here. :raised_hands: